src/Simplex/Messaging/Agent.hs

```diff
@@ -2080,43 +2080,45 @@ data ACKd = ACKd | ACKPending
 -- and sometimes ACK has to be sent from the consumer.
 processSMPTransmissions :: AgentClient -> ServerTransmissionBatch SMPVersion ErrorType BrokerMsg -> AM' ()
 processSMPTransmissions c@AgentClient {subQ} (tSess@(_, srv, _), _v, sessId, ts) = do
-  upConnIds <- newTVarIO []
-  forM_ ts $ \(entId, t) -> case t of
+  connIds <- fmap catMaybes . forM (L.toList ts) $ \(entId, t) -> case t of
     STEvent msgOrErr ->
-      withRcvConn entId $ \rq@RcvQueue {connId} conn -> case msgOrErr of
-        Right msg -> processSMP rq conn (toConnData conn) msg
-        Left e -> lift $ notifyErr connId e
+      withRcvConn entId $ \rq@RcvQueue {connId} conn ->
+        Nothing <$ case msgOrErr of
+          Right msg -> processSMP rq conn (toConnData conn) msg
+          Left e -> lift $ notifyErr connId e
     STResponse (Cmd SRecipient cmd) respOrErr ->
       withRcvConn entId $ \rq conn -> case cmd of
         -- TODO process expired responses to ACK and DEL
         SMP.SUB -> case respOrErr of
-          Right SMP.OK -> processSubOk rq upConnIds
+          Right SMP.OK -> processSubOk rq
           Right msg@SMP.MSG {} -> do
-            processSubOk rq upConnIds
-            processSMP rq conn (toConnData conn) msg
-          Right r -> processSubErr rq $ unexpectedResponse r
-          Left e -> unless (temporaryClientError e) $ processSubErr rq e -- timeout/network was already reported
-        _ -> pure ()
-    STResponse {} -> pure () -- TODO process expired responses to sent messages
+            cId_ <- processSubOk rq
+            cId_ <$ processSMP rq conn (toConnData conn) msg
```

Comment:
the problem here is that if processSMP throws exception (ExceptT or IO) then it won't return cId_ - it will be lost in the handler